### ISSUE TITLE
Question about the parameters

### ISSUE THREAD

**Issue by JinchuanZ** [manigalati/usad#3]
Hello, can I ask why you set the batch_size=7919, and whether the z_size means latent space or not? Did you use the down-sampling rate mentioned in the article USAD? Thanks for your answer.

**Comment by manigalati** [comment]
Hello! The unusual value for the batch size aimed only to speed up the training process, I left it like that but you should be free to change it affecting only the training time. The z_size is actually the multiplication between the window size = 12 and the latent size = 10, and it indeed corresponds to the actual dimension of the latent space. I did not implement the experiments regarding the down-sampling, sorry.

**Comment by JinchuanZ** [comment]
OK, thanks for your help.

**Comment by JinchuanZ** [comment]
Hello, I have another question. I notice that you use the windows_normal_test+windows_attack to be the test_loader, but in the SWaT dataset, it has Normal/Attack label and you drop it, and not all the windows_attack data is abnormal. Thanks for your answer.

**Comment by manigalati** [comment]
Hello again :)
This is definitely an interesting point. The truth is that I reimplemented the code by myself without a strict supervision of the author of the paper, and since he sent me only a simple scratch of his implementation with dataloaders already implemented, I didn't double check the dataset and the labels. So you can be right, if that is the case in the next days I will correct the mistake. Let me know if you try before and get better results.

**Comment by JinchuanZ** [comment]
Hi, I have just read the paper and studied your code, then I found out the question. Until now I still not try to do it. I'm looking forward to your correction and I will try to do it now.

**Comment by severous** [comment]
I  also notice this issue. and I try to do it.but the permence of model is very pool.I am confused about it.

**Comment by severous** [comment]
My implementation is [here](https://github.com/severous/USAD/tree/master)

**Comment by mariazuluaga** [comment]
Hi @severous what do you mean by permence?

**Comment by meihuameii** [comment]
I have the same problem why all attack windows are abnormal?

**Comment by JulienAu** [comment]
> 
> 
> I have the same problem why all attack windows are abnormal?

This problem has been solved in this implementation. 
https://github.com/robustml-eurecom/usad

Thanks

**Comment by meihuameii** [comment]
> > I have the same problem why all attack windows are abnormal?
> 
> This problem has been solved in this implementation.
> https://github.com/robustml-eurecom/usad
> 
> Thanks

Hello, thanks for your reply!  I use the new code to do anomaly detection but I get bad results in f1_score.  It' performance is as good as the original paper.

**Comment by mariazuluaga** [comment]
Hi, 
Could you please provide us with some more details regarding your issue? Which dataset? 
> 
> 
> > > I have the same problem why all attack windows are abnormal?
> > 
> > 
> > This problem has been solved in this implementation.
> > https://github.com/robustml-eurecom/usad
> > Thanks
> 
> Hello, thanks for your reply! I use the new code to do anomaly detection but I get bad results in f1_score. It' performance is not as good as the original paper.

**Comment by meihuameii** [comment]
I have solved this problem！The original paper selected the best F1_score, I  use this way to achieve a similar result in SWaT. Thanks!

**Comment by meihuameii** [comment]
> Hi,
> Could you please provide us with some more details regarding your issue? Which dataset?
> 
> > > > I have the same problem why all attack windows are abnormal?
> > > 
> > > 
> > > This problem has been solved in this implementation.
> > > https://github.com/robustml-eurecom/usad
> > > Thanks
> > 
> > 
> > Hello, thanks for your reply! I use the new code to do anomaly detection but I get bad results in f1_score. It' performance is not as good as the original paper.

I have solved this problem！The original paper selected the best F1_score, I use this way to achieve a similar result in SWaT. Thanks!

**Comment by soemthlng** [comment]
@meihuameii Hello, can you explain how to achieve a similar result in SWaT?

**Comment by hanhuili** [comment]
> > Hi,
> > Could you please provide us with some more details regarding your issue? Which dataset?
> > > > > I have the same problem why all attack windows are abnormal?
> > > > 
> > > > 
> > > > This problem has been solved in this implementation.
> > > > https://github.com/robustml-eurecom/usad
> > > > Thanks
> > > 
> > > 
> > > Hello, thanks for your reply! I use the new code to do anomaly detection but I get bad results in f1_score. It' performance is not as good as the original paper.
> 
> I have solved this problem！The original paper selected the best F1_score, I use this way to achieve a similar result in SWaT. Thanks!

Yet I select the best F1_score, it is about 74%, which is still far below that in the paper

**Comment by finloop** [comment]
> @meihuameii Hello, can you explain how to achieve a similar result in SWaT?

Hi,
I've stumbled upon the same problem and found that increasing threshold can achieve similar results. Here are parameters and scores that I used on one of the runs: 
```
window_size = 20 
BATCH_SIZE =  1000
N_EPOCHS = 50
hidden_size = 20
```
![image](https://user-images.githubusercontent.com/38153367/135798180-5ee31b0f-d8ec-4687-9bfc-35e002a29468.png)

Then I increased `threshold` to `10` and found the predictions with: 

```python
threshold = 10
y_pred_ = np.zeros(y_pred.shape[0])
y_pred_[y_pred >= threshold] = 1
```

Classification report:
```
              precision    recall  f1-score   support

         0.0       0.95      0.99      0.97    394613
         1.0       0.88      0.63      0.74     55286

    accuracy                           0.94    449899
   macro avg       0.92      0.81      0.85    449899
weighted avg       0.94      0.94      0.94    449899
```

**Comment by soemthlng** [comment]
> > @meihuameii Hello, can you explain how to achieve a similar result in SWaT?
> 
> Hi, I've stumbled upon the same problem and found that increasing threshold can achieve similar results. Here are parameters and scores that I used on one of the runs:
> 
> ```
> window_size = 20 
> BATCH_SIZE =  1000
> N_EPOCHS = 50
> hidden_size = 20
> ```
> 
> ![image](https://user-images.githubusercontent.com/38153367/135798180-5ee31b0f-d8ec-4687-9bfc-35e002a29468.png)
> 
> Then I increased `threshold` to `10` and found the predictions with:
> 
> ```python
> threshold = 10
> y_pred_ = np.zeros(y_pred.shape[0])
> y_pred_[y_pred >= threshold] = 1
> ```
> 
> Classification report:
> 
> ```
>               precision    recall  f1-score   support
> 
>          0.0       0.95      0.99      0.97    394613
>          1.0       0.88      0.63      0.74     55286
> 
>     accuracy                           0.94    449899
>    macro avg       0.92      0.81      0.85    449899
> weighted avg       0.94      0.94      0.94    449899
> ```

Hi,
Thank you for sharing your research. Here is my best score and parameters.
I used StandardScaler for preprocessing and I used y_test `y_test=np.concatenate([np.zeros(windows_normal_test.shape[0]), np.ones(windows_attack.shape[0])])` instead of 'y_test = [1.0 if (np.sum(window) > 0) else 0 for window in windows_labels ]'

`window_size  = 8
BATCH_SIZE =  7919
N_EPOCHS = 200
hidden_size = 20`

I checked this score using 'sklearn.metrics'
recall 0.8068351296145237
precision 0.8836299019011222
f1 0.8434881918763455

**Comment by finloop** [comment]
> and did you mean use threshold and prediction in this way?

@soemthlng Yes, I increased `threshold` to trade of recall for f1 (or so I thought, I'm still learning). In my case the score before increasing the threshold was similar to:
```
              precision    recall  f1-score   support

         0.0       0.95      0.72      0.82    394901
         1.0       0.26      0.72      0.38     55006

    accuracy                           0.72    449907
   macro avg       0.61      0.72      0.60    449907
weighted avg       0.86      0.72      0.76    449907
```

How did you select `threshold`?

**Comment by oubashibaika** [comment]
> Hi @severous what do you mean by permence?

Hello, in the loss function of the original paper, ||x||2 represents L2-norm(square root of sum of squares of elements inside x). However, in usad.py, the loss function defined is MES, that is, `loss2 = 1/n*torch.mean((batch-w2)**2)-(1-1/n)*torch.mean((batch-w3)**2)`. Is my understanding wrong? Or is there a problem with the code itself？ Thank you！

**Comment by oubashibaika** [comment]
> Hi, Could you please provide us with some more details regarding your issue? Which dataset?
> 
> > > > I have the same problem why all attack windows are abnormal?
> > > 
> > > 
> > > This problem has been solved in this implementation.
> > > https://github.com/robustml-eurecom/usad
> > > Thanks
> > 
> > 
> > Hello, thanks for your reply! I use the new code to do anomaly detection but I get bad results in f1_score. It' performance is not as good as the original paper.

Hello, have you reproduced the results from the original paper (F1 result without point-adjust on the SWAT dataset is 0.7917). I've been very distressed recently that I can't reproduce the results from my original paper. In my code, BATCH_ SIZE = 1024, N_ EPOCHS = 100, hidden_ Size = 20, window_ Size=10. Standard Scaler () or MinMaxScaler () were used for data preprocessing, and downsampling was used in the code, with a downsampling rate of 5. The best results from these methods were only about 0.74. Do you have any skills to make the results similar to those of the original paper? I'd love to wait for your reply! Thank you very much.

**Comment by dadodimauro** [comment]
Hello,
Has anyone managed to get the results of the paper? if so how?

Thanks!

**Comment by hisong182** [comment]
> y_test = [1.0 if (np.sum(window) > 0) else 0 for window in windows_labels ]

Hello, sorry to bother you, may I ask why you set the threshold to 0, after I set the result 1.0 acc value is 0

**Comment by finloop** [comment]
> Hello, sorry to bother you, may I ask why you set the threshold to 0, after I set the result 1.0 acc value is 0
@hisong182 

A window is considered anomalous if it contains at least one anomaly. Checkout my notebook: https://github.com/finloop/usad/blob/dev/USAD.ipynb

**Comment by hisong182** [comment]
> A window is considered anomalous if it contains at least one anomaly. Checkout my notebook:

Thank you for your answer, before I set the wrong data normalization to MinMaxScaler, after the change the accuracy increased a lot

**Comment by Donek1** [comment]
Thanks. How to get the best result on wadi dataset?

**Comment by Jhonzhang** [comment]
> > Hi, Could you please provide us with some more details regarding your issue? Which dataset?
> > > > > I have the same problem why all attack windows are abnormal?
> > > > 
> > > > 
> > > > This problem has been solved in this implementation.
> > > > https://github.com/robustml-eurecom/usad
> > > > Thanks
> > > 
> > > 
> > > Hello, thanks for your reply! I use the new code to do anomaly detection but I get bad results in f1_score. It' performance is not as good as the original paper.
> 
> Hello, have you reproduced the results from the original paper (F1 result without point-adjust on the SWAT dataset is 0.7917). I've been very distressed recently that I can't reproduce the results from my original paper. In my code, BATCH_ SIZE = 1024, N_ EPOCHS = 100, hidden_ Size = 20, window_ Size=10. Standard Scaler () or MinMaxScaler () were used for data preprocessing, and downsampling was used in the code, with a downsampling rate of 5. The best results from these methods were only about 0.74. Do you have any skills to make the results similar to those of the original paper? I'd love to wait for your reply! Thank you very much.

I have the same questions too. I will check out the new codes which was uploaded by @finloop  https://github.com/finloop/usad/blob/dev/USAD.ipynb

**Comment by Jhonzhang** [comment]
> @soemthlng
> 
> > I think 1.0 class's f1-score is what we want.
> 
> Yes It is what we want.
> 
> > Are you using 0.0 class as true value?
> 
> What do you mean by `true value`?. Class `0.0` is what we consider as normal data. Class `1.0` - anomalies.

I think we should calculate both cases separately, right? And then calculate the average.
